File: packages/kotlin/src/main/kotlin/me/really/crypto/Pbkdf2.kt
// SPDX-FileCopyrightText: Copyright © 2026 ReallyMe LLC. All rights reserved
//
// SPDX-License-Identifier: Apache-2.0

package me.really.crypto

import org.bouncycastle.crypto.digests.SHA256Digest
import org.bouncycastle.crypto.digests.SHA512Digest
import org.bouncycastle.crypto.generators.PKCS5S2ParametersGenerator
import org.bouncycastle.crypto.params.KeyParameter

public object ReallyMePbkdf2 {
    public const val MIN_INPUT_LENGTH: Int = 1
    public const val MAX_INPUT_LENGTH: Int = 4096
    /** Minimum work factor accepted by public PBKDF2 derivation routes. */
    public const val MIN_ITERATIONS: UInt = 100_000u

    /** Maximum work factor accepted before CPU-expensive provider dispatch. */
    public const val MAX_ITERATIONS: UInt = 10_000_000u
    public const val MIN_OUTPUT_LENGTH: Int = 1
    public const val MAX_OUTPUT_LENGTH: Int = 4096

    public fun deriveHmacSha256(
        password: ByteArray,
        salt: ByteArray,
        iterations: UInt,
        outputLength: Int,
    ): ByteArray =
        derive(ProviderAlgorithm.PBKDF2_HMAC_SHA256, password, salt, iterations, outputLength)

    public fun deriveHmacSha512(
        password: ByteArray,
        salt: ByteArray,
        iterations: UInt,
        outputLength: Int,
    ): ByteArray =
        derive(ProviderAlgorithm.PBKDF2_HMAC_SHA512, password, salt, iterations, outputLength)

    private fun derive(
        algorithm: ProviderAlgorithm,
        password: ByteArray,
        salt: ByteArray,
        iterations: UInt,
        outputLength: Int,
    ): ByteArray {
        val providerIterations = validate(password, salt, iterations, outputLength)
        val generator = when (algorithm) {
            ProviderAlgorithm.PBKDF2_HMAC_SHA256 -> PKCS5S2ParametersGenerator(SHA256Digest())
            ProviderAlgorithm.PBKDF2_HMAC_SHA512 -> PKCS5S2ParametersGenerator(SHA512Digest())
        }
        generator.init(password, salt, providerIterations)

        val bitLength = outputLength.checkedMultiply(Byte.SIZE_BITS)
        val parameters = generator.generateDerivedParameters(bitLength)
        if (parameters !is KeyParameter) {
            throw ReallyMeCryptoException.ProviderFailure()
        }
        val derived = parameters.key
        return try {
            derived.copyOf()
        } finally {
            derived.fill(0)
        }
    }

    private fun validate(
        password: ByteArray,
        salt: ByteArray,
        iterations: UInt,
        outputLength: Int,
    ): Int {
        if (password.size !in MIN_INPUT_LENGTH..MAX_INPUT_LENGTH ||
            salt.size !in MIN_INPUT_LENGTH..MAX_INPUT_LENGTH ||
            outputLength !in MIN_OUTPUT_LENGTH..MAX_OUTPUT_LENGTH
        ) {
            throw ReallyMeCryptoException.InvalidInput()
        }
        return checkedIterationCount(iterations)
    }

    internal fun checkedIterationCount(iterations: UInt): Int {
        if (iterations < MIN_ITERATIONS || iterations > MAX_ITERATIONS) {
            throw ReallyMeCryptoException.InvalidInput()
        }
        // The policy range also proves this conversion cannot wrap negative.
        return iterations.toInt()
    }

    private fun Int.checkedMultiply(multiplier: Int): Int {
        val result = this.toLong() * multiplier.toLong()
        if (result > Int.MAX_VALUE) {
            throw ReallyMeCryptoException.InvalidInput()
        }
        return result.toInt()
    }

    private enum class ProviderAlgorithm {
        PBKDF2_HMAC_SHA256,
        PBKDF2_HMAC_SHA512,
    }
}

File: packages/swift/Sources/ReallyMeCrypto/Pbkdf2.swift
// SPDX-FileCopyrightText: Copyright © 2026 ReallyMe LLC. All rights reserved
//
// SPDX-License-Identifier: Apache-2.0

public enum ReallyMePbkdf2 {
    public static let minInputLength = 1
    public static let maxInputLength = 4096
    /// Minimum work factor accepted by public PBKDF2 derivation routes.
    public static let minIterations: UInt32 = 100_000
    /// Maximum work factor accepted before CPU-expensive provider dispatch.
    public static let maxIterations: UInt32 = 10_000_000
    public static let minOutputLength = 1
    public static let maxOutputLength = 4096

    public static func deriveHmacSha256(
        password: [UInt8],
        salt: [UInt8],
        iterations: UInt32,
        outputLength: Int
    ) throws -> [UInt8] {
        try derive(
            password: password,
            salt: salt,
            iterations: iterations,
            outputLength: outputLength,
            hashLength: ReallyMeHmac.sha256TagLength,
            authenticate: ReallyMeHmac.authenticateSha256
        )
    }

    public static func deriveHmacSha512(
        password: [UInt8],
        salt: [UInt8],
        iterations: UInt32,
        outputLength: Int
    ) throws -> [UInt8] {
        try derive(
            password: password,
            salt: salt,
            iterations: iterations,
            outputLength: outputLength,
            hashLength: ReallyMeHmac.sha512TagLength,
            authenticate: ReallyMeHmac.authenticateSha512
        )
    }

    static func derive(
        password: [UInt8],
        salt: [UInt8],
        iterations: UInt32,
        outputLength: Int,
        hashLength: Int,
        authenticate: (_ key: [UInt8], _ message: [UInt8]) throws -> [UInt8],
        clear: (_ bytes: inout [UInt8]) -> Void = ReallyMeCryptoMemory.bestEffortClear
    ) throws -> [UInt8] {
        try validate(password: password, salt: salt, iterations: iterations, outputLength: outputLength)
        guard hashLength > 0 else {
            throw ReallyMeCryptoError.invalidInput
        }

        let adjustedLength = outputLength.addingReportingOverflow(hashLength - 1)
        guard !adjustedLength.overflow else {
            throw ReallyMeCryptoError.invalidInput
        }
        let blockCount = adjustedLength.partialValue / hashLength
        guard blockCount <= Int(UInt32.max) else {
            throw ReallyMeCryptoError.invalidInput
        }

        var derived = [UInt8]()
        derived.reserveCapacity(outputLength)

        do {
            for blockIndex in 1...blockCount {
                guard let blockIndex32 = UInt32(exactly: blockIndex) else {
                    throw ReallyMeCryptoError.invalidInput
                }
                var blockInput = salt
                blockInput.append(UInt8((blockIndex32 >> 24) & 0xff))
                blockInput.append(UInt8((blockIndex32 >> 16) & 0xff))
                blockInput.append(UInt8((blockIndex32 >> 8) & 0xff))
                blockInput.append(UInt8(blockIndex32 & 0xff))

                do {
                    var previous = try authenticate(password, blockInput)
                    guard previous.count == hashLength else {
                        clear(&previous)
                        throw ReallyMeCryptoError.providerFailure
                    }
                    var block = previous
                    defer {
                        clear(&blockInput)
                        clear(&previous)
                        clear(&block)
                    }

                    if iterations > 1 {
                        for _ in 1..<iterations {
                            var next = try authenticate(password, previous)
                            guard next.count == hashLength else {
                                clear(&next)
                                throw ReallyMeCryptoError.providerFailure
                            }
                            clear(&previous)
                            previous = next
                            for index in block.indices {
                                block[index] ^= previous[index]
                            }
                        }
                    }

                    guard derived.count <= outputLength else {
                        throw ReallyMeCryptoError.providerFailure
                    }
                    let remaining = outputLength - derived.count
                    derived.append(contentsOf: block.prefix(remaining))
                } catch {
                    clear(&blockInput)
                    throw error
                }
            }
        } catch {
            // A provider failure after an earlier block must not leave the
            // accumulated derived key to ordinary managed-memory reclamation.
            clear(&derived)
            throw error
        }

        return derived
    }

    private static func validate(
        password: [UInt8],
        salt: [UInt8],
        iterations: UInt32,
        outputLength: Int
    ) throws {
        guard (minInputLength...maxInputLength).contains(password.count),
              (minInputLength...maxInputLength).contains(salt.count),
              (minIterations...maxIterations).contains(iterations),
              (minOutputLength...maxOutputLength).contains(outputLength)
        else {
            throw ReallyMeCryptoError.invalidInput
        }
    }
}
